processing/src/main/java/io/druid/query/groupby/epinephelinae/StreamingMergeSortedGrouper.java

```diff
@@ -0,0 +1,387 @@
+      buffer = bufferSupplier.get();
+      maxSlotNum = buffer.capacity() / recordSize;
+      Preconditions.checkState(
+          maxSlotNum > 2,
+          "Buffer[%s] should be large enough to store at least three records[%s]",
+          buffer.capacity(),
+          recordSize
+      );
+
+      reset();
+      initialized = true;
+    }
+  }
+
+  @Override
+  public boolean isInitialized()
+  {
+    return initialized;
+  }
+
+  @Override
+  public AggregateResult aggregate(KeyType key, int notUsed)
+  {
+    return aggregate(key);
+  }
+
+  @Override
+  public AggregateResult aggregate(KeyType key)
+  {
+    try {
+      final ByteBuffer keyBuffer = keySerde.toByteBuffer(key);
+
+      if (keyBuffer.remaining() != keySize) {
+        throw new IAE(
+            "keySerde.toByteBuffer(key).remaining[%s] != keySerde.keySize[%s], buffer was the wrong size?!",
+            keyBuffer.remaining(),
+            keySize
+        );
+      }
+
+      final int prevRecordOffset = curWriteIndex * recordSize;
+      if (curWriteIndex == -1 || !keyEquals(keyBuffer, buffer, prevRecordOffset)) {
+        initNewSlot(keyBuffer);
+      }
+
+      final int curRecordOffset = curWriteIndex * recordSize;
+      for (int i = 0; i < aggregatorOffsets.length; i++) {
+        aggregators[i].aggregate(buffer, curRecordOffset + aggregatorOffsets[i]);
+      }
+
+      return AggregateResult.ok();
+    }
+    catch (RuntimeException e) {
+      finished = true;
+      throw e;
+    }
+  }
+
+  private boolean keyEquals(ByteBuffer curKeyBuffer, ByteBuffer buffer, int bufferOffset)
+  {
+    int i = 0;
+    for (; i + Long.BYTES <= keySize; i += Long.BYTES) {
+      if (curKeyBuffer.getLong(i) != buffer.getLong(bufferOffset + i)) {
+        return false;
+      }
+    }
+
+    if (i + Integer.BYTES <= keySize) {
+      if (curKeyBuffer.getInt(i) != buffer.getInt(bufferOffset + i)) {
+        return false;
+      }
+      i += Integer.BYTES;
+    }
+
+    for (; i < keySize; i++) {
+      if (curKeyBuffer.get(i) != buffer.get(bufferOffset + i)) {
+        return false;
+      }
+    }
+
+    return true;
+  }
+
+  private void initNewSlot(ByteBuffer newKey)
+  {
+    increaseWriteIndex();
+
+    final int recordOffset = recordSize * curWriteIndex;
+    buffer.position(recordOffset);
+    buffer.put(newKey);
+
+    for (int i = 0; i < aggregators.length; i++) {
+      aggregators[i].init(buffer, recordOffset + aggregatorOffsets[i]);
+    }
+  }
+
+  /**
+   * Wait for {@link #nextReadIndex} to be moved if necessary and move {@link #curWriteIndex}.  {@link #nextReadIndex}
+   * is checked in while loops instead of waiting using a lock to avoid frequent thread park.
+   */
+  private void increaseWriteIndex()
+  {
+    if (curWriteIndex == maxSlotNum - 1) {
+      final long startLoopAt = System.currentTimeMillis();
+      while ((nextReadIndex == -1 || nextReadIndex == 0) && !Thread.interrupted()) {
+        if (System.currentTimeMillis() - startLoopAt > timeoutMs) {
```

Comment:
Spin-looping definitely doesn't make sense beyond 50 microseconds. Start calling Thread.yield() after that.

Reply:
It doesn't account for time corrections and GC pauses. It's more robust to increment "millisecond counter" every time we see any change compared to the prev. value

Reply:
Added yield(). 

> It's more robust to increment "millisecond counter" every time we see any change compared to the prev. value

The timeout here is used to prevent the writing thread from being blocked if the reading thread cannot finish reading due to some errors as noted at https://github.com/druid-io/druid/pull/4704/files#diff-8bb57c81d0c016c3203d4b3af9b741a4R48. Incrementing counter on value changes doesn't solve this problem.

Reply:
Yes, but if a 10-second GC pause hits, this thead may decide that the producer thread has "failed" although it was just frozen for 10 seconds as well as this thread.

Reply:
Made timeout configurable which is set by the timeout of QueryContext.